core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,754 @@
+    slot: Slot,
+    id: &Pubkey,
+    state: VotorMessageType,
+    keypair: &Keypair,
+) {
+    // prepare the packet to send and sign it
+    {
+        let pkt = MockVotePacketHeader::from_bytes_mut(packet_buf);
+        pkt.slot_number = slot;
+        pkt.sender = *id.as_array();
+        pkt.signature = [0; SIGNATURE_BYTES];
+        pkt.state = state as u64;
+    }
+    let signature = keypair.sign_message(&packet_buf[SIGNATURE_BYTES..MOCK_VOTE_HEADER_SIZE]);
+    {
+        let pkt = MockVotePacketHeader::from_bytes_mut(packet_buf);
+        pkt.signature = *signature.as_array();
+    }
+}
+
+// Pubkey for the account that is used to control the test via on-chain state
+mod control_pubkey {
+    solana_pubkey::declare_id!("9PsiyXopc2M9DMEmsEeafNHHHAUmPKe9mHYgrk6fHPyx");
+}
+
+/// Actual on-chain state that controls the mock alpenglow test
+#[derive(Deserialize, Debug)]
+#[repr(C)]
+pub(crate) struct TestConfig {
+    _version: u8,            // This is part of Record program header
+    _authority: [u8; 32],    // This is part of Record program header
+    test_interval_slots: u8, // 0 here means test is disabled
+    verify_signatures: bool,
+    packet_size: u16,
+    _future_use: [u8; 16],
+}
+
+///Parse an account's content, keeping it in cache.
+pub(crate) fn get_test_config_from_account<T: DeserializeOwned>(bank: &Bank) -> Option<T> {
+    let data = bank
+        .accounts()
+        .accounts_db
+        .load_account_with(&bank.ancestors, &control_pubkey::ID, |_| true)?
+        .0;
+    data.deserialize_data().ok()
+}
+
+fn report_collected_votes(peers: HashMap<Pubkey, PeerData>, total_staked: Stake, slot: Slot) {
+    trace!("Reporting statistics for slot {}", slot);
+    let mut total_voted_stake: [Stake; 4] = [0; 4];
+    let mut total_voted_nodes: [usize; 4] = [0; 4];
+    let mut total_delay_ms = [0u128; 4];
+    for (pubkey, peer_data) in peers.iter() {
+        for i in 0..4 {
+            let Some(rel_toa) = peer_data.relative_toa[i] else {
+                continue;
+            };
+            total_voted_stake[i] += peer_data.stake;
+            total_voted_nodes[i] += 1;
+            total_delay_ms[i] +=
+                (rel_toa.as_millis().clamp(0, 800) as u128) * peer_data.stake as u128;
+        }
+    }
+
+    let mut stake_weighted_delay = [0f64; 4];
+    let mut percent_collected = [0f64; 4];
+
+    for i in 0..4 {
+        if total_voted_stake[i] > 0 {
+            stake_weighted_delay[i] = total_delay_ms[i] as f64 / total_voted_stake[i] as f64;
+        }
+        percent_collected[i] = 100.0 * total_voted_stake[i] as f64 / total_staked as f64;
+
+        info!(
+            "{:?}: got {} % of total stake collected, stake-weighted delay is {}ms",
+            VotorMessageType::try_from(i as u64).unwrap(), // this unwrap is ok since i is in static range
+            percent_collected[i],
+            stake_weighted_delay[i]
+        );
+    }
+    datapoint_info!(
+        "mock_alpenglow",
```

Comment:
Hmm, does it help if we put slot somewhere in datapoint_info! ?

Reply:
Can you elaborate on what that would allow us to do?

Reply:
Yeah you are right it would not hurt to log also slot number.